### DISCUSSION TITLE
Implementing my own logic for sending email for verification - token: Missing email token claim.... error

### DISCUSSION THREAD

**Discussion by WingsDevelopment** [pocketbase/pocketbase#2021] [Q&A]
Hello,
I am trying to implement my custom logic for verifing email.
This is what I have so far:
```

await authRepository.register(name, email, password, passwordConfirm);

await authRepository.authWithPassword(email, password);
const token = await authRepository.getToken();

await SendEmail(email, name, token);
```
This is implementation of functions:

```
	authWithPassword = async (email: string, password: string) => {
		const result = await this.pocketBase
			.collection(myCollections.users)
			.authWithPassword(email, password);
		return result;
	};

	getToken = async () => {
		const token = await this.pocketBase.authStore.token;
		return token;
	};
```

this are tokens:
{
  tokenSent: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2Nzk1NjUwOTcsImlkIjoidXQzbnhkNnZhbXl5dnhwIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.SqgOCUVQ50xvJ12JCKQnbM1EYMr5F7Y1QZHx21XXrGQ'
}
{
  tokenReceived: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2Nzk1NjUwOTcsImlkIjoidXQzbnhkNnZhbXl5dnhwIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.SqgOCUVQ50xvJ12JCKQnbM1EYMr5F7Y1QZHx21XXrGQ'
}

and then i just tried to call:

```
const token = String(url.searchParams.get('token') ?? '');
await VerifyUserEmail(getAuthDeps(locals), token);
```

```
VerifyUserEmail= async (verificationToken: string) => {
		console.log({ tokenReceived: verificationToken });
		await this.pocketBase.collection(myCollections.users).confirmVerification(verificationToken);
	};
```

but i am getting an error 

![image](https://user-images.githubusercontent.com/24592343/223986757-5d09015d-bf32-4954-959b-8048c11bb8bd.png)

What am I missing here ?

**Comment by ganigeorgiev** [comment]
`confirmVerification()` accepts only Record Verification tokens. You are pasing a Record Auth token.

I don't understand what do you mean by `tokenSent` and `tokenReceived`.

The Verification token is part of the "Call-to-action" link in the email sent with `requestVerification()`. You can extract the token after the user click on the url (to change the email url, update the "Action url" input in "Admin UI > Settings > Mail settings > Default Verification Email Template").

**Comment by WingsDevelopment** [comment]
How can i get Record Verification token ? 

tokenSent and tokenReceived:
Token sent is token that is sent in email to be href of a confirm button href = baseurl/api/user/verify-email?token=${token}
When user clicks that button, he calls api with that token in params.. that is tokenReceived.

I am sending emails with nodemailer. And I also want to send this verification mail trough nodemailer aswel.
Is this way even possible or I have to use requestVerification() from PB ?

**Reply by ganigeorgiev** [reply]
The auth and the verification token are 2 different tokens and they are not interchangeable.

You have to call `requestVerfication()`.

**Reply by WingsDevelopment** [reply]
This might be a little off topic, 
I am trying to add my email settings in dashboard but i get error
![image](https://user-images.githubusercontent.com/24592343/224024160-35c74fa1-0ce8-4827-854c-1f1270f96af1.png)

But, whatever i try to change for example just adding simple '1' in app name i get same error 
![image](https://user-images.githubusercontent.com/24592343/224023979-375800a3-f060-40f8-b88b-a3736af15ef7.png)

Do you have any idea why?
Creating new records works normally, and application is able to do crud, login etc..

**Reply by ganigeorgiev** [reply]
Hm, this looks similar to https://github.com/pocketbase/pocketbase/issues/1902.

Are you using Docker? If yes, make sure that the settings encryption env is 32 chars string or remove the --encryptionEnv flag.

You can also start PocketBase with `--debug` for more detailed error messages and/or check the logs in the Admin UI > Logs.

**Reply by WingsDevelopment** [reply]
Yes I am on docker. 
Oh i didn't see 'Include requests by admins' checkbox.
The error is :
{
  "errorDetails": "crypto/aes: invalid key size 7",
  "errorMessage": "An error occurred while submitting the form."
}

I will try removing --encryptionEnv flag.

**Reply by WingsDevelopment** [reply]
--encryptionEnv did help, everything works now. thanks!
